platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/ModuleComponentGraphResolveStateFactory.java

```diff
@@ -35,6 +37,11 @@ public ModuleComponentGraphResolveStateFactory(ComponentIdGenerator idFactory, A
     }
 
     public ModuleComponentGraphResolveState stateFor(ModuleComponentResolveMetadata metadata) {
+        if (metadata instanceof IvyModuleResolveMetadata) {
+            IvyModuleResolveMetadata ivyMetadata = (IvyModuleResolveMetadata) metadata;
+            return new DefaultIvyComponentGraphResolveState(idGenerator.nextComponentId(), ivyMetadata, attributeDesugaring, idGenerator);
+        }
+
         return new DefaultModuleComponentGraphResolveState(idGenerator.nextComponentId(), metadata, attributeDesugaring, idGenerator);
```

Comment:
🤔 This seems backwards from an OOP point of view.

Why not:

```
public ModuleComponentGraphResolveState stateFor(ModuleComponentResolveMetadata metadata) {
   return metadata.buildResolveState(idGenerator, attributeDesugaring);
}
```

Reply:
The intention is that the State types are mutable/stateful and track all the data for a given component while the Metadata types are fully immutable and track a subset of data for a given component. 

Each component should have one state type, but that state type may track multiple metadata types as the state changes (we may learn about a component's variants through different network requests/operations compared to its artifacts). In that case we would have a metadata for the artifacts, a metadata for the variants, and some base metadata that describes the identity of the components and some other basic stuff that we know up-front about it. 

Currently for external module components (and local components -- but I am working to change that), the Metadata type (`ModuleComponentResolveMetadata` in this file) is stateful and has a lot of logic that the State type that wraps it _should_ do.

The point that I'm trying to make here is that the State type should be "in charge" of the Metadata type. It feels wrong to me for the metadata type to construct a state type for itself.

Reply:
OK, I see your point.  Can you add some of that to the Javadoc on this factory?